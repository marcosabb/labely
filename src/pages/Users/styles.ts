import styled from 'styled-components/native'
import Icon from 'react-native-vector-icons/MaterialCommunityIcons'
import { theme } from 'styled-tools'

import { FlatList } from 'react-native'
import { User } from '../../contexts/users'

export const Container = styled.View`
  flex: 1;
  background-color: ${theme('colors.background')};
`

export const HeaderLeft = styled.View`
  margin-left: ${theme('spacing.default')};
`

export const HeaderRight = styled.View`
  margin-right: ${theme('spacing.default')};
`

export const GithubIcon = styled(Icon).attrs(({ theme }) => ({
  name: 'github',
  color: theme.colors.black,
  size: 32
}))``

export const List = styled(FlatList as new () => FlatList<User>).attrs({
  contentContainerStyle: {
    paddingTop: 8,
    paddingHorizontal: 16
  }
})``
