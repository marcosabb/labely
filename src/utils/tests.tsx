import React from 'react'
import { render, RenderOptions } from '@testing-library/react-native'
import { ThemeProvider } from 'styled-components/native'

import theme from 'styles/theme'

interface Props {
  children: React.ReactNode
}

function ThemeWrapper({ children }: Props) {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>
}

const customRender = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, 'queries'>
) => render(ui, { wrapper: ThemeWrapper, ...options })

export * from '@testing-library/react-native'

export { customRender as render }
