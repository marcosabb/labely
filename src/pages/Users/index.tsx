import React, { useCallback, useEffect } from 'react'
import { useNavigation } from '@react-navigation/native'

import { User, useUsers } from '../../contexts/users'

import Button from '../../components/Button'
import Item from '../../components/Item'

import { Container, HeaderLeft, HeaderRight, GithubIcon, List } from './styles'

interface RenderProps {
  item: User
}

export default function Users() {
  const { navigate, setOptions } = useNavigation()
  const { users, deleteUser } = useUsers()

  const handleNavigate = useCallback(() => {
    navigate('User', { isCreating: true })
  }, [navigate])

  useEffect(() => {
    setOptions({
      headerLeft: () => (
        <HeaderLeft>
          <GithubIcon />
        </HeaderLeft>
      ),

      headerRight: () => (
        <HeaderRight>
          <Button size='small' testID='header-button' onPress={handleNavigate}>
            Adicionar novo
          </Button>
        </HeaderRight>
      )
    })
  }, [handleNavigate, setOptions])

  const handleRemove = useCallback(
    async (id: number) => {
      await deleteUser(id)
    },
    [deleteUser]
  )

  function keyExtractor({ id }: User) {
    return String(id)
  }

  function renderItem({
    item: { id, name, login, avatar_url, company, location, starred }
  }: RenderProps) {
    function formatStarred() {
      if (starred) {
        if (starred >= 30) return `${starred}+`

        return starred
      }
    }

    const tags = [
      { id: 1, icon: 'business', value: company },
      { id: 2, icon: 'location-on', value: location },
      {
        id: 3,
        icon: 'star',
        value: formatStarred()
      }
    ]

    return (
      <Item
        title={name}
        description={`@${login}`}
        avatar={avatar_url}
        tags={tags}
        onRemove={() => handleRemove(id)}
      />
    )
  }

  return (
    <Container>
      <List data={users} keyExtractor={keyExtractor} renderItem={renderItem} />
    </Container>
  )
}
