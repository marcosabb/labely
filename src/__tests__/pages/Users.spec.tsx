import React from 'react'

import { render } from '../../utils/tests'

import Users from '../../pages/Users'

const mockedNavigate = jest.fn()

jest.mock('@react-navigation/native', () => ({
  useNavigation: () => ({
    navigate: mockedNavigate,
    setOptions: () => ({
      headerRight: () => {}
    })
  })
}))

jest.mock('../../contexts/users', () => ({
  useUsers: () => ({
    users: [
      {
        id: 56012991,
        name: 'Marcos Borges',
        avatar_url: 'https://avatars.githubusercontent.com/u/56012991?v=4',
        login: 'marcosabb',
        company: 'Incentivar.io',
        location: 'BrasÃ­lia, Brazil',
        starred: 30
      }
    ]
  })
}))

describe('Users page', () => {
  it('should match snapshot', () => {
    const component = render(<Users />)

    expect(component).toMatchSnapshot()
  })

  it('should render items if has users', () => {
    const { getByTestId } = render(<Users />)

    const title = getByTestId('item-title')

    expect(title).toBeTruthy()
  })
})
