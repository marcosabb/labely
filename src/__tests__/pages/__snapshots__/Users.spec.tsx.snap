// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Users page should match snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#e5e5e5",
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
      },
    ]
  }
>
  <RCTScrollView
    contentContainerStyle={
      Object {
        "paddingHorizontal": 16,
        "paddingTop": 8,
      }
    }
    data={
      Array [
        Object {
          "avatar_url": "https://avatars.githubusercontent.com/u/56012991?v=4",
          "company": "Incentivar.io",
          "id": 56012991,
          "location": "Brasília, Brazil",
          "login": "marcosabb",
          "name": "Marcos Borges",
          "starred": 30,
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    style={
      Array [
        Object {},
      ]
    }
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "backgroundColor": "#fff",
              "borderRadius": 8,
              "marginBottom": 8,
              "opacity": 1,
              "paddingBottom": 32,
              "paddingLeft": 24,
              "paddingRight": 24,
              "paddingTop": 32,
            }
          }
          testID="item-container"
        >
          <View
            style={
              Array [
                Object {
                  "flexDirection": "row",
                  "marginBottom": 16,
                },
              ]
            }
          >
            <Image
              source={
                Object {
                  "uri": "https://avatars.githubusercontent.com/u/56012991?v=4",
                }
              }
              style={
                Array [
                  Object {
                    "borderRadius": 32,
                    "height": 64,
                    "marginRight": 16,
                    "width": 64,
                  },
                ]
              }
            />
            <View
              style={
                Array [
                  Object {
                    "flexBasis": 0,
                    "flexGrow": 1,
                    "flexShrink": 1,
                    "justifyContent": "center",
                  },
                ]
              }
            >
              <View
                style={
                  Array [
                    Object {
                      "flexDirection": "row",
                    },
                  ]
                }
              >
                <Text
                  numberOfLines={1}
                  style={
                    Array [
                      Object {
                        "color": "#000",
                        "fontFamily": "Mulish-Bold",
                        "fontSize": 16,
                        "marginBottom": 10,
                        "marginRight": 10,
                        "maxWidth": "50%",
                      },
                    ]
                  }
                  testID="item-title"
                >
                  Marcos Borges
                </Text>
                <Text
                  allowFontScaling={false}
                  style={
                    Array [
                      Object {
                        "color": undefined,
                        "fontSize": 12,
                      },
                      Array [
                        Object {
                          "color": "#000",
                          "flexBasis": 0,
                          "flexGrow": 1,
                          "flexShrink": 1,
                          "fontSize": 24,
                        },
                      ],
                      Object {
                        "fontFamily": "Material Icons",
                        "fontStyle": "normal",
                        "fontWeight": "normal",
                      },
                      Object {},
                    ]
                  }
                >
                  
                </Text>
                <View
                  accessible={true}
                  focusable={true}
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "#e8e8e8",
                      "borderRadius": 12,
                      "height": 24,
                      "justifyContent": "center",
                      "opacity": 1,
                      "width": 24,
                    }
                  }
                  testID="delete-button"
                >
                  <Text
                    allowFontScaling={false}
                    style={
                      Array [
                        Object {
                          "color": undefined,
                          "fontSize": 12,
                        },
                        Array [
                          Object {
                            "color": "#000",
                            "fontSize": 16,
                          },
                        ],
                        Object {
                          "fontFamily": "Material Icons",
                          "fontStyle": "normal",
                          "fontWeight": "normal",
                        },
                        Object {},
                      ]
                    }
                  >
                    
                  </Text>
                </View>
              </View>
              <Text
                numberOfLines={2}
                style={
                  Array [
                    Object {
                      "color": "#000",
                      "fontFamily": "Mulish-Regular",
                      "fontSize": 14,
                    },
                  ]
                }
              >
                @marcosabb
              </Text>
            </View>
          </View>
          <View
            style={
              Array [
                Object {
                  "flexDirection": "row",
                },
              ]
            }
          >
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "marginRight": 16,
                  },
                ]
              }
            >
              <Text
                allowFontScaling={false}
                style={
                  Array [
                    Object {
                      "color": undefined,
                      "fontSize": 12,
                    },
                    Array [
                      Object {
                        "color": "#e5e5e5",
                        "fontSize": 16,
                        "marginRight": 4,
                      },
                    ],
                    Object {
                      "fontFamily": "Material Icons",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    Object {},
                  ]
                }
              >
                
              </Text>
              <Text
                style={
                  Array [
                    Object {
                      "color": "#7e7e7e",
                      "fontFamily": "Mulish-Regular",
                      "fontSize": 12,
                    },
                  ]
                }
              >
                Incentivar.io
              </Text>
            </View>
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "marginRight": 16,
                  },
                ]
              }
            >
              <Text
                allowFontScaling={false}
                style={
                  Array [
                    Object {
                      "color": undefined,
                      "fontSize": 12,
                    },
                    Array [
                      Object {
                        "color": "#e5e5e5",
                        "fontSize": 16,
                        "marginRight": 4,
                      },
                    ],
                    Object {
                      "fontFamily": "Material Icons",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    Object {},
                  ]
                }
              >
                
              </Text>
              <Text
                style={
                  Array [
                    Object {
                      "color": "#7e7e7e",
                      "fontFamily": "Mulish-Regular",
                      "fontSize": 12,
                    },
                  ]
                }
              >
                Brasília, Brazil
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
